processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -179,133 +164,108 @@ public V getValue(Object key)
-    {
-      @Override
-      protected ByteBuffer initialValue()
-      {
-        return theBuffer.asReadOnlyBuffer();
-      }
-    };
-
-    this.cacheable = false;
-    this.cachedValues = new ThreadLocal<>();
-  }
-
-  /**
-   * Creates a copy of the given indexed with the given cache size
-   * The resulting copy must be closed to release resources used by the cache
-   */
-  GenericIndexed(GenericIndexed<T> other, final int maxBytes)
-  {
-    this.theBuffer = other.theBuffer;
-    this.strategy = other.strategy;
-    this.allowReverseLookup = other.allowReverseLookup;
-    this.size = other.size;
-    this.indexOffset = other.indexOffset;
-    this.valuesOffset = other.valuesOffset;
-    this.cachedBuffer = other.cachedBuffer;
-
-    this.cachedValues = new ThreadLocal<SizedLRUMap<Integer, T>>() {
-      @Override
-      protected SizedLRUMap<Integer, T> initialValue()
-      {
-        log.debug("Allocating column cache of max size[%d]", maxBytes);
-        return new SizedLRUMap<>(INITIAL_CACHE_CAPACITY, maxBytes);
-      }
-    };
-
-    this.cacheable = strategy instanceof CacheableObjectStrategy;
-  }
-
-  @Override
-  public Class<? extends T> getClazz()
-  {
-    return strategy.getClazz();
+    bufferIndexed = new BufferIndexed();
   }
 
-  @Override
-  public int size()
+  public class BufferIndexed implements Indexed<T>
   {
-    return size;
-  }
+    int lastReadSize;
 
-  @Override
-  public T get(int index)
-  {
-    if (index < 0) {
-      throw new IAE("Index[%s] < 0", index);
+    @Override
+    public Class<? extends T> getClazz()
+    {
+      return strategy.getClazz();
     }
-    if (index >= size) {
-      throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+
+    @Override
+    public int size()
+    {
+      return size;
     }
 
-    if(cacheable) {
-      final T cached = cachedValues.get().getValue(index);
-      if (cached != null) {
-        return cached;
-      }
+    @Override
+    public T get(final int index)
+    {
+      return _get(theBuffer.asReadOnlyBuffer(), index);
     }
 
-    // using a cached copy of the buffer instead of making a read-only copy every time get() is called is faster
-    final ByteBuffer copyBuffer = this.cachedBuffer.get();
+    protected T _get(final ByteBuffer copyBuffer, final int index)
+    {
+      if (index < 0) {
+        throw new IAE("Index[%s] < 0", index);
+      }
+      if (index >= size) {
+        throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+      }
 
-    final int startOffset;
-    final int endOffset;
+      final int startOffset;
+      final int endOffset;
 
-    if (index == 0) {
-      startOffset = 4;
-      endOffset = copyBuffer.getInt(indexOffset);
-    } else {
-      copyBuffer.position(indexOffset + ((index - 1) * 4));
-      startOffset = copyBuffer.getInt() + 4;
-      endOffset = copyBuffer.getInt();
-    }
+      if (index == 0) {
+        startOffset = 4;
```

Comment:
I would rather limit the changes to what is required to address the problem and not mix in unrelated changes, in order to make it clear this commit did not change the underlying logic.